x-pack/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.ts

```diff
@@ -9,41 +9,39 @@ import {
   SERVICE_ENVIRONMENT,
   SERVICE_NAME,
 } from '../../../common/elasticsearch_fieldnames';
-import {
-  Connection,
-  ConnectionNode,
-  ServiceConnectionNode,
-} from '../../../common/service_map';
+import { Connection, ConnectionNode } from '../../../common/service_map';
 import { Setup } from '../helpers/setup_request';
 import { fetchServicePathsFromTraceIds } from './fetch_service_paths_from_trace_ids';
 
-export function getConnections(
-  paths?: ConnectionNode[][],
-  serviceName?: string,
-  environment?: string
-) {
+export function getConnections({
+  paths,
+  serviceName,
+  environment,
+}: {
+  paths: ConnectionNode[][] | undefined;
+  serviceName: string | undefined;
+  environment: string | undefined;
+}) {
   if (!paths) {
     return [];
   }
 
   if (serviceName || environment) {
     paths = paths.filter((path) => {
       return path.some((node) => {
-        let matches = true;
-        if (serviceName) {
-          matches =
-            matches &&
-            SERVICE_NAME in node &&
-            (node as ServiceConnectionNode)[SERVICE_NAME] === serviceName;
+        if (serviceName && node[SERVICE_NAME] !== serviceName) {
+          return false;
         }
-        if (environment && environment !== ENVIRONMENT_NOT_DEFINED.value) {
-          matches =
-            matches &&
-            SERVICE_ENVIRONMENT in node &&
-            (node as ServiceConnectionNode)[SERVICE_ENVIRONMENT] ===
-              environment;
+
+        if (!environment) {
+          return true;
+        }
+
+        if (environment === ENVIRONMENT_NOT_DEFINED.value) {
+          return node['span.type'] !== 'external' && !node[SERVICE_ENVIRONMENT];
```

Comment:
Why are we checking for `span.type == 'external'`?

Reply:
Good catch

Reply:
@cauemarcondes Could this have been caught by a test? 🤔

Reply:
I don't think so, because when ENVIRONMENT_NOT_DEFINED is selected, we filter services which don't have `service.environment`. The `node['span.type'] !== 'external'` filter is kind of useless in that case because it would filter the path based on the service node information, not the external node.

For example, a scenario like this will never happen, where `service.environment=production` and `environment === ENVIRONMENT_NOT_DEFINED.value` is true, and in that case `node['span.type'] !== 'external'` **would be used**.
```js
[
    {
      'service.environment': 'production',
      'service.name': 'opbeans-ruby',
      'agent.name': 'ruby',
    },
    {
      'service.environment': 'production',
      'service.name': 'opbeans-python',
      'agent.name': 'python',
    },
    {
      'service.environment': 'production',
      'service.name': 'opbeans-ruby',
      'agent.name': 'ruby',
    },
    {
      'service.environment': 'production',
      'service.name': 'opbeans-python',
      'agent.name': 'python',
    },
    {
      'span.subtype': 'http',
      'span.destination.service.resource': 'opbeans-node:3000',
      'span.type': 'external',
    },
  ],
```

But like said early, when ENVIRONMENT_NOT_DEFINED is selected, this would be an example response:
```js
[
    {
      'service.environment': null,
      'service.name': 'opbeans-ruby',
      'agent.name': 'ruby',
    },
    {
      'service.environment': null,
      'service.name': 'opbeans-python',
      'agent.name': 'python',
    },
    {
      'service.environment': null,
      'service.name': 'opbeans-ruby',
      'agent.name': 'ruby',
    },
    {
      'service.environment': null,
      'service.name': 'opbeans-python',
      'agent.name': 'python',
    },
    {
      'span.subtype': 'http',
      'span.destination.service.resource': 'opbeans-node:3000',
      'span.type': 'external',
    },
  ],
```
And in that case, the `external` node wound't even be checked because the first node doesn't have `service.environment` defined.